File: commons/javafx3d/src/main/java/com/cedarsoft/commons/javafx/threed/TransformUtil.java
package com.cedarsoft.commons.javafx.threed;

import java.util.Collection;

import javax.annotation.Nonnull;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.collect.ImmutableList;

import javafx.beans.value.ChangeListener;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.scene.Node;
import javafx.scene.transform.NonInvertibleTransformException;
import javafx.scene.transform.Rotate;
import javafx.scene.transform.Transform;
import javafx.scene.transform.Translate;

/**
 * A utility class around {@link Transform}s.
 */
public class TransformUtil {
  @Nonnull 
  private static final Logger LOG = LoggerFactory.getLogger(TransformUtil.class.getName());

  private TransformUtil() {
    // utility class
  }

  public static void installInverseRotate(@Nonnull Node referenceNode, @Nonnull Node... nodes) {
    installInverseRotate(referenceNode, ImmutableList.copyOf(nodes));
  }

  /**
   * Whenever the reference node is rotated (via a {@link Rotate} transformation) all other nodes a rotated with the opposite angle.
   */
  public static void installInverseRotate(@Nonnull Node referenceNode, @Nonnull Collection<? extends Node> nodes) {
    // Remove all old rotate transformations.
    nodes.forEach(TransformUtil::removeAllRotations);

    // Add a ChangeListener for all current rotations set on the reference node.
    ChangeListener<Number> rotateListener = (observable, oldValue, newValue) -> invertRotate(referenceNode, nodes);
    referenceNode.getTransforms().stream()
      .filter(transform -> transform instanceof Rotate)
      .forEach(rotate -> ((Rotate) rotate).angleProperty().addListener(rotateListener));

    // Add a ChangeListener for all new rotations set on the reference node and
    // delete the old ChangeListener for all rotations removed from the reference node.
    referenceNode.getTransforms().addListener((ListChangeListener<Transform>) change -> {
      while (change.next()) {
        change.getRemoved().stream().filter(transform -> transform instanceof Rotate).forEach(rotate -> {
          ((Rotate) rotate).angleProperty().removeListener(rotateListener);
        });
        change.getAddedSubList().stream().filter(transform -> transform instanceof Rotate).forEach(rotate -> {
          ((Rotate) rotate).angleProperty().addListener(rotateListener);
        });
      }
      invertRotate(referenceNode, nodes);
    });

    invertRotate(referenceNode, nodes);
  }

  /**
   * Inverts all transforms of type {@link Rotate} from the reference node and applies them to each node of the list.
   */
  private static void invertRotate(@Nonnull Node referenceNode, @Nonnull Collection<? extends Node> nodes) {
    ObservableList<Transform> transforms = referenceNode.getTransforms();
    if (transforms.isEmpty()) {
      return;
    }
    Transform combinedRotate = new Translate(0, 0, 0);
    for (Transform transform : transforms) {
      if (!(transform instanceof Rotate)) {
        continue;
      }
      combinedRotate = combinedRotate.createConcatenation(transform);
    }
    try {
      Transform inverseTransform = combinedRotate.createInverse();
      nodes.forEach(node -> node.getTransforms().setAll(inverseTransform));
    } catch (NonInvertibleTransformException e) {
      LOG.error("failed to invert rotation", e);
    }
  }

  /**
   * Remove all instances of {@link Rotate} from the list of transforms of the given node.
   */
  public static void removeAllRotations(@Nonnull Node node) {
    node.getTransforms().removeIf(transform -> transform instanceof Rotate);
  }

  /**
   * Binds all translations and rotations from a target node to a source node.
   */
  public static void bindAll(@Nonnull Node source, @Nonnull Node target) {
    target.translateXProperty().bind(source.translateXProperty());
    target.translateYProperty().bind(source.translateYProperty());
    target.translateZProperty().bind(source.translateZProperty());

    target.scaleXProperty().bind(source.scaleXProperty());
    target.scaleYProperty().bind(source.scaleYProperty());
    target.scaleZProperty().bind(source.scaleZProperty());

    target.rotateProperty().bind(source.rotateProperty());
    target.rotationAxisProperty().bind(source.rotationAxisProperty());

    source.getTransforms().addListener((ListChangeListener<Transform>) change -> target.getTransforms().setAll(source.getTransforms()));
    target.getTransforms().setAll(source.getTransforms());
  }
}

File: commons/javafx3d/src/main/kotlin/com/cedarsoft/commons/javafx/threed/TransformUtil.kt
package com.cedarsoft.commons.javafx.threed

import com.google.common.collect.ImmutableList
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.collections.ListChangeListener
import javafx.scene.Node
import javafx.scene.transform.NonInvertibleTransformException
import javafx.scene.transform.Rotate
import javafx.scene.transform.Transform
import javafx.scene.transform.Translate
import org.slf4j.LoggerFactory

/**
 * A utility class around [Transform]s.
 */
object TransformUtil {
  private val LOG = LoggerFactory.getLogger(TransformUtil::class.java.name)

  @JvmStatic
  fun installInverseRotate(referenceNode: Node, vararg nodes: Node) {
    installInverseRotate(referenceNode, ImmutableList.copyOf(nodes))
  }

  /**
   * Whenever the reference node is rotated (via a [Rotate] transformation) all other nodes a rotated with the opposite angle.
   */
  @JvmStatic
  fun installInverseRotate(referenceNode: Node, nodes: Collection<Node>) {
    // Remove all old rotate transformations.
    nodes.forEach {
      removeAllRotations(it)
    }

    // Add a ChangeListener for all current rotations set on the reference node.
    val rotateListener = ChangeListener { observable: ObservableValue<out Number?>?, oldValue: Number?, newValue: Number? -> invertRotate(referenceNode, nodes) }
    referenceNode.transforms.stream().filter { transform: Transform? -> transform is Rotate }.forEach { rotate: Transform -> (rotate as Rotate).angleProperty().addListener(rotateListener) }

    // Add a ChangeListener for all new rotations set on the reference node and
    // delete the old ChangeListener for all rotations removed from the reference node.
    referenceNode.transforms.addListener { change: ListChangeListener.Change<out Transform> ->
      while (change.next()) {
        change.removed.stream().filter { transform: Transform? -> transform is Rotate }.forEach { rotate: Transform -> (rotate as Rotate).angleProperty().removeListener(rotateListener) }
        change.addedSubList.stream().filter { transform: Transform? -> transform is Rotate }.forEach { rotate: Transform -> (rotate as Rotate).angleProperty().addListener(rotateListener) }
      }
      invertRotate(referenceNode, nodes)
    }
    invertRotate(referenceNode, nodes)
  }

  /**
   * Inverts all transforms of type [Rotate] from the reference node and applies them to each node of the list.
   */
  private fun invertRotate(referenceNode: Node, nodes: Collection<Node>) {
    val transforms = referenceNode.transforms
    if (transforms.isEmpty()) {
      return
    }
    var combinedRotate: Transform = Translate(0.0, 0.0, 0.0)
    for (transform in transforms) {
      if (transform !is Rotate) {
        continue
      }
      combinedRotate = combinedRotate.createConcatenation(transform)
    }
    try {
      val inverseTransform = combinedRotate.createInverse()
      nodes.forEach { node: Node -> node.transforms.setAll(inverseTransform) }
    } catch (e: NonInvertibleTransformException) {
      LOG.error("failed to invert rotation", e)
    }
  }

  /**
   * Remove all instances of [Rotate] from the list of transforms of the given node.
   */
  @JvmStatic
  fun removeAllRotations(node: Node) {
    node.transforms.removeIf { transform: Transform? -> transform is Rotate }
  }

  /**
   * Binds all translations and rotations from a target node to a source node.
   */
  @JvmStatic
  fun bindAll(source: Node, target: Node) {
    target.translateXProperty().bind(source.translateXProperty())
    target.translateYProperty().bind(source.translateYProperty())
    target.translateZProperty().bind(source.translateZProperty())
    target.scaleXProperty().bind(source.scaleXProperty())
    target.scaleYProperty().bind(source.scaleYProperty())
    target.scaleZProperty().bind(source.scaleZProperty())
    target.rotateProperty().bind(source.rotateProperty())
    target.rotationAxisProperty().bind(source.rotationAxisProperty())
    source.transforms.addListener { change: ListChangeListener.Change<out Transform?>? -> target.transforms.setAll(source.transforms) }
    target.transforms.setAll(source.transforms)
  }
}
